src/include/migraphx/op/scatternd_op.hpp

```diff
@@ -28,44 +28,82 @@
-        auto k         = inputs.at(1).lens().back();
-        auto ind_lens  = inputs.at(1).lens();
-        auto upd_lens  = inputs.back().lens();
-        auto data_lens = inputs.front().lens();
-        if(k > r)
-            MIGRAPHX_THROW("ScatterND: index of size " + std::to_string(k) +
-                           " is too large for tensor of rank " + std::to_string(r));
-        if(not(std::equal(ind_lens.begin(), ind_lens.begin() + q - 1, upd_lens.begin()) and
-               std::equal(data_lens.begin() + k, data_lens.end(), upd_lens.begin() + q - 1)))
-            MIGRAPHX_THROW("ScatterND: incorrect update shape. update.lens != indices.lens[0:q-1] "
-                           "++ data.lens[k:r-1]");
-        auto s = inputs.front();
-        if(s.broadcasted())
+        check_shapes{inputs, *this, true}.has(3);
+
+        if(migraphx::none_of(inputs, [](auto v) { return v.dynamic(); }))
         {
-            return {s.type(), s.lens()};
+            auto r         = inputs.front().ndim();
+            auto q         = inputs.at(1).ndim();
+            auto k         = inputs.at(1).lens().back();
+            auto ind_lens  = inputs.at(1).lens();
+            auto upd_lens  = inputs.back().lens();
+            auto data_lens = inputs.front().lens();
+
+            if(q + r - k - 1 != upd_lens.size())
+                MIGRAPHX_THROW("ScatterND:  ranks of inputs don't match. " + std::to_string(q) +
+                               " + " + std::to_string(r) + " - " + std::to_string(k) +
+                               " - 1 != " + std::to_string(upd_lens.size()));
+            if(k > r)
+                MIGRAPHX_THROW("ScatterND: index of size " + std::to_string(k) +
+                               " is too large for tensor of rank " + std::to_string(r));
+            if(not(std::equal(ind_lens.begin(), ind_lens.begin() + q - 1, upd_lens.begin()) and
+                   std::equal(data_lens.begin() + k, data_lens.end(), upd_lens.begin() + q - 1)))
+                MIGRAPHX_THROW(
+                    "ScatterND: incorrect update shape. update.lens != indices.lens[0:q-1] "
+                    "++ data.lens[k:r-1]");
+            auto s = inputs.front();
+            if(s.broadcasted())
+            {
+                return {s.type(), s.lens()};
+            }
+            else
+            {
+                return s.with_lens(s.lens());
+            }
         }
         else
         {
+            // One or more inputs are dynamic; shape validation can't be done yet.  Putting off
+            // validation until 3 static shapes are received means that if any inputs are dynamic
+            // shapes, the validation will happen at evaluation time.
+            // TODO: gathernd op. handles the situation differently by checking dynamic shape inputs
+            // in compute_shape() but requiring the last dimension of index (k) to be fixed, which
+            // allows the checks to take place.  Should the same be done here?
```

Comment:
We should check conditions on ranks (ndims) as static shape here as well. Those conditions remain valid for dynamic shapes as well. 
e.g. q + r - k - 1 != upd_lens.size()

Reply:
I have significantly refactored the dimension checking to make the static and dynamic cases more similar.  Question for @CharlieL7 : will converting static shapes with to_dynamic() before comparing dimensions give wrong results for broadcasted dimensions?